lib/msf/core/exploit/remote/ms_lsarpc.rb

```diff
@@ -0,0 +1,136 @@
+###
+#
+# This mixin provides methods to open, and close policy handles, and to query policy info on the remote SMB server.
+#
+# -*- coding: binary -*-
+
+module Msf
+
+  module Exploit::Remote::MsLsarpc
+
+    include Msf::Exploit::Remote::MsIpc
+
+    class MsLsarpcError < StandardError; end
+    class MsLsarpcConnectionError < MsLsarpcError; end
+    class MsLsarpcAuthenticationError < MsLsarpcError; end
+    class MsLsarpcUnexpectedReplyError < MsLsarpcError; end
+
+    LSA_UUID        = '12345778-1234-abcd-ef00-0123456789ab'.freeze
+    LSA_VERS        = '0.0'.freeze
+    LSARPC_ENDPOINT = RubySMB::Dcerpc::Lsarpc.freeze
+
+    # The currently connected LSARPC pipe
+    attr_reader :lsarpc_pipe
+
+    def map_security_principal_to_string(security_principal)
+      case security_principal
+      when 1
+        'User'
+      when 2
+        'Group'
+      when 3
+        'Domain'
+      when 4
+        'Alias'
+      when 5
+        'Well-Known Group'
+      when 6
+        'Deleted Account'
+      when 7
+        'Invalid'
+      when 8
+        'Unknown'
+      when '9'
+        'Computer'
+      when 10
+        'Label'
+      else
+        'Unknown - Not a valid Security Principal'
+      end
+    end
+
+    def open_policy_2(impersonation_level, security_context_tracking_mode, access_mask)
+      object_attributes = LSARPC_ENDPOINT::LsaprObjectAttributes.new(
+        {
+          len: 24,
+          root_directory: nil,
+          object_name: nil,
+          attributes: 0,
+          security_descriptor: nil,
+          security_quality_of_service: {
+            len: 12,
+            impersonation_level: impersonation_level,
+            security_context_tracking_mode: security_context_tracking_mode,
+            effective_only: 0
+          }
+        }
+      )
+
+      self.lsarpc_pipe.lsar_open_policy_2(
+        system_name: simple.peerhost,
+        object_attributes: object_attributes,
+        access_mask: access_mask
+      )
+    end
+
+    def query_information_policy(policy_handle, information_class)
+      self.lsarpc_pipe.lsar_query_information_policy(
+        policy_handle: policy_handle,
+        information_class: information_class
+      )
+    end
+
+    def lookup_sids(policy_handle, sids_buffer, lookup_level)
+      self.lsarpc_pipe.lsar_lookup_sids(
+        policy_handle: policy_handle,
+        sid_enum_buffer: {
+          num_entries: sids_buffer.count,
+          sid_info: sids_buffer
+        },
```

Comment:
This seems like the type of thing that could be abstracted away into [`#lsar_lookup_sids`](https://github.com/rapid7/ruby_smb/pull/267/files#diff-08ef9e0605f90149beb43c4715cf38aa208d6db58fd1fdf4a3029cff7b77bd22R620). That would decouple the internal structure of the LsaprSidEnumBuffer in the unlikely event that it changed. It seems reasonable that `#lsar_lookup_sids` would construct it itself from an array of sid strings.